### DISCUSSION TITLE
Add device_id to context

### DISCUSSION THREAD

**Discussion by balloob** [home-assistant/architecture#978] [General] [answered]
## Context

Context is an object that indicates what actions belong to one another. It is attached to events and states. It is also passed along to service calls such that any resulting event/state shares the context.

By looking at the history, we note that the first event with a specific context was the originating event that set the things in motion.

Besides an `id`, a context has other fields, all optional:

 - `parent_id`: this is used to link contexts together. If an automation triggers because of a state change, it will start a new context and the `parent_context_id` is set to the context ID of the state change.
 - `user_id`: this is used to indicate the user that is making this action. It is used for both authentication and to attribute a change to a user

With voice we're introducing voice assistants. These are devices that take voice commands that will interact with Home Assistant. It is important that a user is able to track down which device initiated a command. Voice identification might help with this in the future, but it's not a guarantee to always be there nor would we block commands from non-identified users.

We also see events fired from devices or in the case of ESPHome, services called from a device. In both cases it's not clear what the origin is.

## Proposal

Add a new `device_id` to Context to indicate what device started the action.

## Consequences

We add a new optional field `device_id` to `Context` class in `homeassistant/core.py`.

We need to store this data in the database. Since the context is stored flattened in the events and states table, it will require adding a new column to both tables.

ESPHome will need to set `device_id` when creating a context as part of Assist Pipeline or any service call/event fired from ESPHome.

**Comment by balloob** [comment]
Here is an open question: we've been moving away from integrations firing events for things like motion detected in favor of new `event` entities. Do we want to update the old events to include `device_id` in the context?

**Reply by frenck** [reply]
We have discussed this today, and concluded we should enrich anything we can. The more we store, the more we are able to show.

**Comment by BradleyFord** [comment]
Yes, it can only help when debugging

**Comment by luka6000** [comment]
Can this be generalized also to automation not only voice triggers? Looking at automation trace we can see what actions it performed, say call service switch toggle. But looking at that switch logbook we only see state change from service call (sometimes with user name) but not the automation entity that actually triggered that action. So maybe extending context with entity_id would be better. Maybe it's not entity_id or device_id but both? The purpose stays the same: being able to track down source of action.

**Reply by balloob** [reply]
This is already possible. The `automation_triggered` event was the first event with the context, and so Home Assistant can derive who fired it.

**Comment by frenck** [comment, marked as answer]
Discussed this proposal today, and agreed this is OK to move forward.

../Frenck

**Comment by witold-gren** [comment]
I'm not sure if this is related, but I think it's also worth adding `device_id` to `intent_script`, specially when we try to run `action` and we need select some `target.entity_id`..
